tket/src/Transformations/SingleQubitSquash/SingleQubitSquash.hpp

```diff
@@ -0,0 +1,312 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <optional>
+
+#include "Gate/Gate.hpp"
+#include "Gate/GatePtr.hpp"
+#include "Transformations/Transform.hpp"
+#include "Utils/Expression.hpp"
+
+namespace tket {
+
+/**
+ * @brief Squashes single qubit gates using given Squasher
+ *
+ * The template Squasher must define the following interface:
+ *   - bool accepts(OpType): whether the OpType can be added to current squash
+ *   - void append(Gate_ptr): add gate to current squash
+ *   - std::pair<Circuit, Gate_ptr> flush(std::optional<Pauli>):
+ *     return a squashed circuit, optionally use the commutation colour of the
+ *     next gate to return an additional Gate_ptr to be commuted through
+ *     (otherwise, set Gate_ptr = nullptr)
+ *   - void clear(): reset the current squash
+ */
+template <typename Squasher>
+class SingleQubitSquash {
+ private:
+  using Condition = std::optional<std::pair<std::list<VertPort>, unsigned>>;
+
+ public:
+  SingleQubitSquash(const Squasher &squasher, bool reversed = false)
+      : squasher_(squasher), reversed_(reversed), circ_ptr_(nullptr) {}
+
+  // squash an entire circuit, one qubit at a time
+  bool squash(Circuit &circ) {
+    bool success = false;
+    circ_ptr_ = &circ;
+
+    VertexVec inputs = circ.q_inputs();
+    VertexVec outputs = circ.q_outputs();
+    for (unsigned i = 0; i < circ.n_qubits(); ++i) {
+      Edge in = circ.get_nth_out_edge(inputs[i], 0);
+      Edge out = circ.get_nth_in_edge(outputs[i], 0);
+      if (reversed_) {
+        success |= squash_between(out, in);
+      } else {
+        success |= squash_between(in, out);
+      }
+    }
+
+    circ_ptr_ = nullptr;
+    return success;
+  }
+
+  // squash everything between edges in and out
+  bool squash_between(const Edge &in, const Edge &out) {
```

Comment:
I'm refactoring it to use polymorphism instead of templating. This makes the abstract interface more explicit and allows me to move all the implementation into a separate `cpp` file